### PR TITLE
Simplify CMake  CUDNN code

### PR DESCRIPTION
**Pull request by cyyever**
1. Move CUDNN code to seperate module.
2. Merge CUDNN public and private targets into a single private target. There is no need to expose CUDNN dependency.

### CODE REVIEW

**Comment by cyyever** [comment]
@malfet  Can you review it?

File: torch/CMakeLists.txt [hunk 1]
```diff
@@ -132,6 +132,7 @@ if(USE_CUDA)
 
     list(APPEND TORCH_PYTHON_COMPILE_DEFINITIONS USE_CUDA)
     if(USE_CUDNN)
+        list(APPEND TORCH_PYTHON_LINK_LIBRARIES torch::cudnn)
```
Comment (malfet) [inline torch/CMakeLists.txt:135]: `TORCH_PYTHON` should not need to be aware of any CUDA/CUDNN related headers/libraries, as it should interact with the hardware via lib torch device abstractions
```suggestion
```

File: cmake/public/cuda.cmake [hunk 2]
```diff
@@ -305,49 +257,37 @@ set_property(
...
+  endif()
+
+  find_package(CUDNN)
+
+  if(NOT CUDNN_FOUND)
+    message(WARNING
+      "Cannot find cuDNN library. Turning the option off")
+    set(CAFFE2_USE_CUDNN OFF)
+  else()
+    if(CUDNN_VERSION VERSION_LESS "7.0.0")
```
Comment (malfet) [inline cmake/public/cuda.cmake:276]: Please change it to 8 or above
```suggestion
    if(CUDNN_VERSION VERSION_LESS "8.0.0")
```

**Review by malfet — APPROVED** [review]
Looks good to me, but please bump minimum CuDNN version as well as remove CuDNN dependency from `libtorch_python` (it should not make any cudnn calls directly, only via `at::cuda::` abstractions or include any cudnn headers. )

File: cmake/public/cuda.cmake [hunk 3]
```diff
@@ -273,7 +273,7 @@ if(CAFFE2_USE_CUDNN)
       "Cannot find cuDNN library. Turning the option off")
     set(CAFFE2_USE_CUDNN OFF)
   else()
-    if(CUDNN_VERSION VERSION_LESS "7.0.0")
+    if(CUDNN_VERSION VERSION_LESS "8.0.0")
       message(FATAL_ERROR "PyTorch requires cuDNN 7 and above.")
```
Comment (Skylion007) [inline cmake/public/cuda.cmake:277]: @cyyever Nit. Fatal_Error message doesn't match condition.

File: torch/CMakeLists.txt (hunk 1 above)
Comment (malfet) [reply torch/CMakeLists.txt:135]: Hmm, why did you mark this conversation as resolved?

Comment (cyyever) [reply torch/CMakeLists.txt:135]: I committed  your advice before and then I marked it as resolved. But then the build failed, showing that this line is still required. Now the change has been reverted.

Comment (cyyever) [reply torch/CMakeLists.txt:135]: The reason is that while the dependency is private, cudnn headers are exposed in public headers.

**Pull request closed without merging**